lib/db_utils.go

```diff
@@ -10462,83 +10465,123 @@ func DBDeletePostAssociationWithTxn(txn *badger.Txn, snap *Snapshot, association
+	return errors.Wrap(txn.Set(_dbKeyForTransactorNonceEntry(nonce, pkid), []byte{}),
+		"DbPutTransactorNonceEntryWithTxn: Problem setting nonce")
+}
 
-	return nil
+func DbPutTransactorNonceEntry(handle *badger.DB, nonce *DeSoNonce, pkid *PKID) error {
+	return handle.Update(func(txn *badger.Txn) error {
+		return DbPutTransactorNonceEntryWithTxn(txn, nonce, pkid)
+	})
 }
 
-func DbPutNextNonceForPKID(handle *badger.DB, pkid PKID, nextNonce uint64) error {
+func DbDeleteTransactorNonceEntryWithTxn(txn *badger.Txn, nonce *DeSoNonce, pkid *PKID) error {
+	return errors.Wrap(
+		txn.Delete(_dbKeyForTransactorNonceEntry(nonce, pkid)),
+		"DbDeleteTransactorNonceEntryWithTxn: Problem deleting nonce")
+}
 
+func DbDeleteExpiredTransactorNonceEntriesAtBlockHeight(handle *badger.DB, blockHeight uint64) error {
 	return handle.Update(func(txn *badger.Txn) error {
-		return DbPutNextNonceForPKIDWithTxn(txn, pkid, nextNonce)
+		return DbDeleteExpiredTransactorNonceEntriesAtBlockHeightWithTxn(txn, blockHeight)
 	})
 }
 
-func DbDeleteNextNonceForPKIDWithTxn(
-	txn *badger.Txn, pkid PKID) error {
+func DbDeleteExpiredTransactorNonceEntriesAtBlockHeightWithTxn(txn *badger.Txn, blockHeight uint64) error {
 
-	if err := txn.Delete(_dbKeyForPKIDToNextNonce(pkid)); err != nil {
-		return errors.Wrapf(
-			err, "DbDeleteNextNonceForPKIDfoWithTxn: Problem deleting next nonce for: %s ",
-			PkToStringBoth(pkid[:]))
+	startPrefix := _dbPrefixForNonceEntryIndexWithBlockHeight(blockHeight)
+	endPrefix := append([]byte{}, Prefixes.PrefixNoncePKIDIndex...)
+	// Add 33 max bytes for the PKID
+	for ii := 0; ii < 33; ii++ {
+		startPrefix = append(startPrefix, 0xFF)
+	}
+	// And then append the max uint64 to start key
+	startPrefix = append(startPrefix, EncodeUint64(math.MaxUint64)...)
+	opts := badger.DefaultIteratorOptions
+	opts.Reverse = true
+	nodeIterator := txn.NewIterator(opts)
+	defer nodeIterator.Close()
+	keysToDelete := [][]byte{}
+	for nodeIterator.Seek(startPrefix); nodeIterator.ValidForPrefix(endPrefix); nodeIterator.Next() {
+		keysToDelete = append(keysToDelete, nodeIterator.Item().Key())
+	}
+	for _, key := range keysToDelete {
+		if err := DBDeleteWithTxn(txn, nil, key); err != nil {
+			return errors.Wrapf(err,
+				"DbDeleteExpiredTransactorNonceEntriesAtBlockHeightWithTxn: Problem deleting key: %v", key)
+		}
 	}
 	return nil
 }
 
+func DbGetAllTransactorNonceEntries(handle *badger.DB) []*TransactorNonceEntry {
+	keys, _ := EnumerateKeysForPrefix(handle, Prefixes.PrefixNoncePKIDIndex)
+	nonceEntries := []*TransactorNonceEntry{}
+	for _, key := range keys {
+		// Convert key to nonce entry.
+		nonceEntries = append(nonceEntries, TransactorNonceKeyToTransactorNonceEntry(key))
+	}
+	return nonceEntries
+}
+
+func TransactorNonceKeyToTransactorNonceEntry(key []byte) *TransactorNonceEntry {
+	keyWithoutPrefix := key[1:]
+	expirationHeight := DecodeUint64(keyWithoutPrefix[:8])
+	pkid := &PKID{}
+	copy(pkid[:], keyWithoutPrefix[8:41])
```

Comment:
This seems a little finicky to do the manual byte slicing we're doing here. Probably fine.